processing/src/main/java/org/apache/druid/query/filter/InDimFilter.java

```diff
@@ -84,9 +90,11 @@ public InDimFilter(
     Preconditions.checkNotNull(dimension, "dimension can not be null");
     Preconditions.checkArgument(values != null, "values can not be null");
 
-    this.values = new TreeSet<>(Comparators.naturalNullsFirst());
-    for (String value : values) {
-      this.values.add(NullHandling.emptyToNullIfNeeded(value));
+    // The values set can be huge. Try to avoid copying the set if possible.
+    if (values instanceof Set && values.stream().noneMatch(NullHandling::needsEmptyToNull)) {
```

Comment:
If `NullHandling.replaceWithDefault()` returns false, I think you can skip the scan here and just reuse the set

Reply:
Good point. Added a check for it.